src/datasets/arrow_writer.py

```diff
@@ -87,14 +88,18 @@ def __arrow_array__(self, type=None):
         """This function is called when calling pa.array(typed_sequence)"""
         assert type is None, "TypedSequence is supposed to be used with pa.array(typed_sequence, type=None)"
         trying_type = False
-        if type is None and self.try_type:
+        if type is not None:
+            pass
+        elif type is None and self.try_type:
             type = self.try_type
             trying_type = True
         else:
             type = self.type
         try:
             if isinstance(type, _ArrayXDExtensionType):
                 out = pa.ExtensionArray.from_storage(type, pa.array(self.data, type.storage_dtype))
```

Comment:
Here we can also support numpy arrays for extension types as well (Array2D, Array3D etc.):

```suggestion
                storage = pa.array(self.data, type.storage_dtype) if not isinstance(self.data, np.ndarray) else numpy_to_pyarrow_listarray(self.data, type=type.value_type)
                out = pa.ExtensionArray.from_storage(type, storage)
```

We just need to edit `numpy_to_pyarrow_listarray` to have a `type` parameter that is passed at
```python
values = pa.array(arr.flatten(), type=type)
```